packages/context/test/acceptance/class-level-bindings.ts

```diff
@@ -177,6 +185,305 @@ describe('Context bindings - Injecting dependencies of classes', () => {
+      .toClass(Store)
+      .withOptions({x: 2, y: 'b'});
+
+    const store1 = await ctx.get('store1');
+    expect(store1.optionX).to.eql(1);
+    expect(store1.optionY).to.eql('a');
+
+    const store2 = await ctx.get('store2');
+    expect(store2.optionX).to.eql(2);
+    expect(store2.optionY).to.eql('b');
+  });
+
+  it('injects undefined option if no binding is present', async () => {
+    class Store {
+      constructor(
+        // tslint:disable-next-line:no-any
+        @inject.options('x') public option: string | undefined,
+      ) {}
+    }
+
+    const store = await RejectionError.reject(instantiateClass(Store, ctx));
+    expect(store.option).to.be.undefined();
+  });
+
+  it('injects values by tag', () => {
+    class Store {
+      constructor(@inject.tag('store:location') public locations: string[]) {}
+    }
+
+    ctx.bind('store').toClass(Store);
+    ctx
+      .bind('store.locations.sf')
+      .to('San Francisco')
+      .tag('store:location');
+    ctx
+      .bind('store.locations.sj')
+      .to('San Jose')
+      .tag('store:location');
+    const store: Store = ctx.getSync('store');
+    expect(store.locations).to.eql(['San Francisco', 'San Jose']);
+  });
+
+  it('injects values by tag regex', () => {
+    class Store {
+      constructor(
+        @inject.tag(/.+:location:sj/)
+        public locations: string[],
+      ) {}
+    }
+
+    ctx.bind('store').toClass(Store);
+    ctx
+      .bind('store.locations.sf')
+      .to('San Francisco')
+      .tag('store:location:sf');
+    ctx
+      .bind('store.locations.sj')
+      .to('San Jose')
+      .tag('store:location:sj');
+    const store: Store = ctx.getSync('store');
+    expect(store.locations).to.eql(['San Jose']);
+  });
+
+  it('injects empty values by tag if not found', () => {
+    class Store {
+      constructor(@inject.tag('store:location') public locations: string[]) {}
+    }
+
+    ctx.bind('store').toClass(Store);
+    const store: Store = ctx.getSync('store');
+    expect(store.locations).to.eql([]);
+  });
+
+  it('injects values by tag asynchronously', async () => {
+    class Store {
+      constructor(@inject.tag('store:location') public locations: string[]) {}
+    }
+
+    ctx.bind('store').toClass(Store);
+    ctx
+      .bind('store.locations.sf')
+      .to('San Francisco')
+      .tag('store:location');
+    ctx
+      .bind('store.locations.sj')
+      .to(Promise.resolve('San Jose'))
+      .tag('store:location');
+    const store: Store = await ctx.get('store');
+    expect(store.locations).to.eql(['San Francisco', 'San Jose']);
+  });
+
+  it('reports error when @inject.tag rejects a promise', async () => {
+    class Store {
+      constructor(@inject.tag('store:location') public locations: string[]) {}
+    }
+
+    ctx.bind('store').toClass(Store);
+    ctx
+      .bind('store.locations.sf')
+      .to('San Francisco')
+      .tag('store:location');
+    ctx
+      .bind('store.locations.sj')
+      .to(Promise.reject('Bad'))
```

Comment:
Rejecting promises with a non-Error value is considered as anti-pattern AFAIK. Please use a real Error object.